### PR TITLE
GODRIVER-2325 Expose raw server response in relevant errors

### PR DESCRIPTION
**Pull request by benjirewis**
GODRIVER-2325

Adds `Raw` fields to various `mongo` and `driver` errors. Adds integration tests to ensure the raw server reply can be accessed through these fields.

### CODE REVIEW

File: mongo/integration/errors_test.go [hunk 1]
```diff
@@ -322,6 +326,126 @@ func TestErrors(t *testing.T) {
...
+						Times: 1,
+					},
+					Data: mtest.FailPointData{
+						FailCommands: []string{"delete"},
+						ErrorCode:    123,
+					},
+				})
+
+				_, err := mt.Coll.DeleteOne(context.Background(), bson.D{})
+				assert.NotNil(mt, err, "expected DeleteOne error, got nil")
```
Comment (kevinAlbs) [inline mongo/integration/errors_test.go:373]: This suggestion assumes the above suggestion to insert duplicate `_id` is accepted.

I think the error will be a `WriteException` containing a `WriteError`.

```suggestion
				assert.NotNil(mt, err, "expected DeleteOne error, got nil")
				_, ok := err.(mongo.WriteException)
				assert.True(mt, ok, "expected type mongo.WriteException, got %v", reflect.TypeOf(err))
```

File: mongo/errors.go [hunk 2]
```diff
@@ -277,6 +288,7 @@ type WriteError struct {
 	Code    int
 	Message string
 	Details bson.Raw
+	Raw     bsoncore.Document // The original server response containing the error.
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply mongo/errors.go:291]: Ah very true; updated documentation.

File: mongo/errors.go [hunk 3]
```diff
@@ -344,6 +362,7 @@ type WriteConcernError struct {
...
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply mongo/errors.go:365]: Yep! Updated.

File: mongo/errors.go [hunk 4]
```diff
@@ -551,6 +579,18 @@ func (bwe BulkWriteException) HasErrorCodeWithMessage(code int, message string)
...
+// RawResponse returns the original server responses containing the errors.
+func (bwe BulkWriteException) RawResponse() []bsoncore.Document {
+	responses := make([]bsoncore.Document, 0)
+	if bwe.WriteConcernError != nil {
+		responses = append(responses, bwe.WriteConcernError.Raw)
+	}
+	for _, we := range bwe.WriteErrors {
+		responses = append(responses, we.RawResponse()[0])
+	}
+	return responses
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply mongo/errors.go:591]: > Apologies for the contradictory feedback.

No worries at all; I think you're definitely right 🧑‍🔧 . I'm still looking into a better api for `BulkWriteException` but will update in this thread when I find one. Removed the `RawResponse` method from the interface, though.

File: mongo/integration/errors_test.go (hunk 1 above)
Comment (benjirewis) [reply mongo/integration/errors_test.go:373]: Yep! Updated to extract the `WriteError` from the `WriteException`.

File: mongo/integration/errors_test.go [hunk 5]
```diff
@@ -322,6 +326,126 @@ func TestErrors(t *testing.T) {
...
+					Data: mtest.FailPointData{
+						FailCommands: []string{"delete"},
+						WriteConcernError: &mtest.WriteConcernErrorData{
+							Code: 123,
+						},
+					},
+				})
+
+				_, err := mt.Coll.DeleteMany(context.Background(), bson.D{})
+				assert.NotNil(mt, err, "expected DeleteMany error, got nil")
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply mongo/integration/errors_test.go:402]: Done.

File: mongo/integration/errors_test.go [hunk 6]
```diff
@@ -322,6 +326,126 @@ func TestErrors(t *testing.T) {
...
+				// Assert that raw response exists and contains error code 123.
+				raw := se.RawResponse()[0]
+				assert.NotNil(mt, raw, "expected RawResponse, got nil")
+				val, err := raw.LookupErr("code")
+				assert.Nil(mt, err, "expected 'code' field in RawResponse, got %v", raw)
+				code, ok := val.AsInt64OK()
+				assert.True(mt, ok, "expected 'code' to be int64, got %v", val)
+				assert.Equal(mt, code, int64(123), "expected 'code' 123, got %d", code)
+			})
+			mt.Run("WriteError", func(mt *mtest.T) {
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply mongo/integration/errors_test.go:359]: Great idea; switched to duplicate `_id` insertion.

File: mongo/integration/errors_test.go [hunk 7]
```diff
@@ -322,6 +326,126 @@ func TestErrors(t *testing.T) {
...
+						Times: 1,
+					},
+					Data: mtest.FailPointData{
+						FailCommands: []string{"find"},
+						ErrorCode:    123,
+					},
+				})
+
+				res := mt.Coll.FindOne(context.Background(), bson.D{})
+				assert.NotNil(mt, res.Err(), "expected FindOne error, got nil")
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply mongo/integration/errors_test.go:346]: Yep, no longer casting to `ServerError`.

File: mongo/integration/errors_test.go [hunk 8]
```diff
@@ -322,6 +326,126 @@ func TestErrors(t *testing.T) {
...
+						},
+					},
+				})
+
+				models := []mongo.WriteModel{
+					&mongo.InsertOneModel{bson.D{{"x", 1}}},
+					&mongo.DeleteOneModel{bson.D{{"x", 1}}, nil, nil},
+				}
+				_, err := mt.Coll.BulkWrite(context.Background(), models)
+				assert.NotNil(mt, err, "expected BulkWrite error, got nil")
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply mongo/integration/errors_test.go:435]: This test is removed for now, but will do type assertion once I add it back in.

**Review by kevinAlbs — APPROVED** [review]
LGTM. Nicely done!

@WenbinZhu can you review to validate this solves the intended use case?

**Review by WenbinZhu — APPROVED** [review]
Thanks for the change, LGTM! (Really sorry for the delay, my email notification setting routes all emails from non 10gen repos to my personal mailbox and I overlooked the review request email...)

**Comment by benjirewis** [comment]
No worries @WenbinZhu ! Thanks for taking a look.

**Pull request merged**